README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# Sbus
This library communicates with SBUS receivers and servos and is compatible with Arduino and CMake build systems.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Description
SBUS is a bus protocol for receivers to send commands to servos. Unlike PWM, SBUS uses a bus architecture where a single serial line can be connected with up to 16 servos with each receiving a unique command.

The SBUS protocol uses an inverted serial logic with a baud rate of 100000, 8 data bits, even parity, and 2 stop bits. The SBUS packet is 25 bytes long consisting of:
   * Byte[0]: SBUS header, 0x0F
   * Byte[1 -22]: 16 servo channels, 11 bits each
   * Byte[23]
      * Bit 0: channel 17 (0x01)
      * Bit 1: channel 18 (0x02)
      * Bit 2: frame lost (0x04)
      * Bit 3: failsafe activated (0x08)
   * Byte[24]: SBUS footer

Note that lost frame is indicated when a frame is lost between the transmitter and receiver. Failsafe activation typically requires that many frames are lost in a row and indicates that the receiver has moved into failsafe mode. Packets are sent approximately every 10 ms or 20 ms, depending on the system configuration.

A variation on SBUS called "Fast SBUS" has started to be used. This uses a baudrate of 200000 and a quicker update rate.

**Note on CH17 and CH18:** Channel 17 and channel 18 are digital on/off channels. These are not universally available on all SBUS receivers and servos.

FrSky receivers will output a range of 172 - 1811 with channels set to a range of -100% to +100%. Using extended limits of -150% to +150% outputs a range of 0 to 2047, which is the maximum range acheivable with 11 bits of data.

Because SBUS is a digital bus format, it is an excellent means of receiving pilot commands from a transmitter and an SBUS capable receiver. If SBUS servos are used in the aircraft, SBUS is also an excellent means of sending actuator commands - servo commands can often be sent with lower latency and, by only using a single pin to command up to 16 servos, additional microcontroller pins are freed for other uses.

# Inverted Serial
SBUS uses an inverted serial protocol, which is not commonly supported in Arduino. This library is able to use inverted serial for the following microcontrollers:
   * Teensy 3.x
   * Teensy 4.x
   * Teensy LC
   * STM32L496xx
   * STM32L476xx
   * STM32L433xx
   * STM32L432xx
   * ESP32

For all other microcontrollers, you **must** use a serial inverter. If you've modified this library to work with other microcontrollers, please submit a pull request.

# Installation

## Arduino
Simply clone or download and extract the zipped library into your Arduino/libraries folder. The library is added as:

```C++
#include "sbus.h"
```

An example is located in *examples/arduino/sbus_example/sbus_example.ino*. This library is tested with Teensy 3.x, 4.x, and LC devices and should work with other Arduino devices.

## CMake
CMake is used to build this library, which is exported as a library target called *sbus*. The header is added as:

```C++
#include "sbus.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake .. -DMCU=MK66FX1M0
make
```

This will build the library and example executable called *sbus_example*. The example executable source file is located at *examples/cmake/sbus_example.cc*. Notice that the *cmake* command includes a define specifying the microcontroller the code is being compiled for. This is required to correctly configure the code, CPU frequency, and compile/linker options. The available MCUs are:
   * MK64FX512
   * MK66FX1M0
   * MKL26Z64
   * IMXRT1062_T40
   * IMXRT1062_T41
   * IMXRT1062_MMOD

These are known to work with the same packages used in Teensy products. Also switching packages is known to work well, as long as it's only a package change.

The *sbus_example* target creates an executable for communicating with sbus receivers and servos. This target also has a *_hex* for creating the hex file and an *_upload* for using the [Teensy CLI Uploader](https://www.pjrc.com/teensy/loader_cli.html) to flash the Teensy. Instructions for setting up your build environment can be found in our [build-tools repo](https://github.com/bolderflight/build-tools). 

# Namespace
This library is within the namespace *bfs*.

# SbusData
This struct defines SBUS data that can be read and returned by the *SbusRx* object or set and sent by the *SbusTx* object.

## Data Members

**bool lost_frame** Whether a frame has been lost.

**bool failsafe** Whether the receiver has entered failsafe mode or to command servos to enter failsafe mode.

**bool ch17, ch18** State of channel 17 and channel 18.

**static constexpr int8_t NUM_CH = 16** The number of SBUS channels.

**int16_t ch[NUM_CH]** An array of SBUS channel data.

# SbusRx
This class is used for receiving SBUS data from an SBUS capable receiver.

**SbusRx(HardwareSerial &ast;bus)** Creates an *SbusRx* object. A pointer to the *Serial* object corresponding to the serial port used is passed. The RX pin of the serial port will receive SBUS packets.

```C++
bfs::SbusRx sbus(&Serial1);
```

**SbusRx(HardwareSerial &ast;bus, const bool inv)** Creates an *SbusRx* object. A pointer to the *Serial* object corresponding to the serial port used is passed along with a second parameter, *inv*, which sets whether inverted serial is used. If *inv* is true, the signal is the standard inverted SBUS, otherwise it is non-inverted SBUS.

```C++
bfs::SbusRx sbus(&Serial1, false);
```

**SbusRx(HardwareSerial &ast;bus, const bool inv, const bool fast)** Same as the constructor above, but enables selecting the fast SBUS baudrate (200000) if *fast* is true.

**(ESP32 ONLY) SbusRx(HardwareSerial &ast;bus, const int8_t rxpin, const int8_t txpin, const bool inv)** Creates an *SbusRx* object. A pointer to the *Serial* object corresponding to the serial port used is passed along with the RX pin number (*rxpin*), TX pin number (*txpin*), and whether inverted serial is used (*inv*). If *inv* is true, the signal is the standard inverted SBUS, otherwise it is non-inverted SBUS.

**(ESP32 ONLY) SbusRx(HardwareSerial &ast;bus, const int8_t rxpin, const int8_t txpin, const bool inv, const bool fast)** Same as the constructor above, but enables selecting the fast SBUS baudrate (200000) if *fast* is true.

**void Begin()** Initializes SBUS communication.

```C++
sbus.Begin();
```

**bool Read()** Parses SBUS packets, returns true on successfully receiving an SBUS packet.

```C++
if (sbus.Read()) {
  // Do something with the received data
}
```

**SbusData data()** Returns the *SbusData* structure, populated with data from the last received packet.

```C++
if (sbus.Read()) {
  bfs::SbusData data = sbus.data();
}
```

# SbusTx
This class is used for transmitting SBUS data to SBUS capable servos.

**SbusTx(HardwareSerial &ast;bus)** Creates an *SbusTx* object. A pointer to the *Serial* object corresponding to the serial port used is passed. The TX pin of the serial port will receive SBUS packets.

```C++
bfs::SbusTx sbus(&Serial1);
```

**SbusTx(HardwareSerial &ast;bus, const bool inv)** Creates an *SbusTx* object. A pointer to the *Serial* object corresponding to the serial port used is passed along with a second parameter, *inv*, which sets whether inverted serial is used. If *inv* is true, the signal is the standard inverted SBUS, otherwise it is non-inverted SBUS.

```C++
bfs::SbusTx sbus(&Serial1, false);
```

**(ESP32 ONLY) SbusTx(HardwareSerial &ast;bus, const int8_t rxpin, const int8_t txpin, const bool inv)** Creates an *SbusTx* object. A pointer to the *Serial* object corresponding to the serial port used is passed along with the RX pin number (*rxpin*), TX pin number (*txpin*), and whether inverted serial is used (*inv*). If *inv* is true, the signal is the standard inverted SBUS, otherwise it is non-inverted SBUS.

**void Begin()** Initializes SBUS communication.

```C++
sbus.Begin();
```

**void Write()** Writes an SBUS packet. The packet is written immediately, you should regulate timing of sending packets to servos to maintain a frequency of approximately 100 Hz or 50 Hz, depending on the setup of the SBUS system.

```C++
sbus.Write();
```

**void data(const SbusData &data)** Sets the SBUS data, which will be transmitted on the next *Write* method.

```C++
bfs::SbusData data;
data.ch[0] = 900;
sbus.data(data);
```

**SbusData data()** Returns the SBUS data buffered in the *SbusTx* object, which will be transmitted on the next *Write* method.

```C++
bfs::SbusData data = sbus.data();
```


CHANGELOG.md
# Changelog

## v8.1.4
- Fixed a bug where trying to write SBUS packets with Teensy 3.1/3.2 would hang the processor by de-referencing a null pointer

## v8.1.3
- Updated core to v3.1.3

## v8.1.2
- Updated core to v3.1.2

## v8.1.1
- Updated core to support MMOD

## v8.1.0
- Added option for fast SBUS (200000 baud)

## v8.0.1
- Enabling ESP32 to use non-inverted SBUS

## v8.0.0
- Fixed bug in SbusRx timing where the last packet waited for the start of the next packet before returning true
- Added option to specify a non-inverted signal for cases where that is a hardware option
- Removed std::array dependency to facilitate use on Arduino AVR
- Updated ESP32 implementation for greater consistency to other processors
- Implemented data struct to ease reading and setting SBUS values without needing std::array

## v7.0.0
- Using std::array for passing around SBUS data and commands

## v6.0.3
- Cleaned up folder structure after merge
- Pulling in mcu-support repo for CMake builds
- Fixed pointer constness for SbusRx

## v6.0.2
- Removed the tools folder to fix Arduino linting

## v6.0.1
- Fixing library.properties version number

## v6.0.0
- Merging CMake and Arduino SBUS libraries. Version 6 is the next available version number common between the two.

## v2.1.2
- Added Embedded Template Library support for AVR boards.

## v2.1.1
- Added CH17 and 18 support for read and write
- Added capability to write lost_frame or failsafe

## v2.0.1
- Added back in support for SBUS2 footers

## v2.0.0
- Updated to match our [SBUS](https://github.com/bolderflight/sbus) library for flight software
- Updated license to MIT

## v1.0.1
- Updated license to GPLV3.

## v1.0.0
- Updated to Arduino 1.5 format and setting a baseline release here.


src/sbus.cpp
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "sbus.h"  // NOLINT
#if defined(ARDUINO)
#include <Arduino.h>
#else
#include <cstddef>
#include <cstdint>
#include "core/core.h"
#endif

namespace bfs {

void SbusRx::Begin() {
  if (fast_) {
    baud_ = 200000;
  } else {
    baud_ = 100000;
  }
  /* Start the bus */
  /* Teensy 3.0 || Teensy 3.1/3.2 */
  #if defined(__MK20DX128__) || defined(__MK20DX256__)
  if (inv_) {
    uart_->begin(baud_, SERIAL_8E1_RXINV_TXINV);
  } else {
    uart_->begin(baud_, SERIAL_8E1);
  }
  /*
  * Teensy 3.5 || Teensy 3.6 ||
  * Teensy LC  || Teensy 4.0/4.1 ||
  * Teensy 4.0 Beta
  */
  #elif defined(__MK64FX512__) || defined(__MK66FX1M0__) || \
        defined(__MKL26Z64__)  || defined(__IMXRT1062__) || \
        defined(__IMXRT1052__)
  if (inv_){
    uart_->begin(baud_, SERIAL_8E2_RXINV_TXINV);
  } else {
    uart_->begin(baud_, SERIAL_8E2);
  }
  /* STM32L4 */
  #elif defined(STM32L496xx) || defined(STM32L476xx) || \
        defined(STM32L433xx) || defined(STM32L432xx)
  if (inv_) {
    uart_->begin(baud_, SERIAL_8E2 | 0xC000ul);
  } else {
    uart_->begin(baud_, SERIAL_8E2);
  }
  /* ESP32 */
  #elif defined(ESP32)
  uart_->begin(baud_, SERIAL_8E2, rxpin_, txpin_, inv_);
  /* Everything else, with a hardware inverter */
  #else
  uart_->begin(baud_, SERIAL_8E2);
  #endif
  /* flush the bus */
  uart_->flush();
}

bool SbusRx::Read() {
  /* Read through all available packets to get the newest */
  new_data_ = false;
  do {
    if (Parse()) {
      new_data_ = true;
    }
  } while (uart_->available());
  return new_data_;
}

bool SbusRx::Parse() {
  /* Parse messages */
  while (uart_->available()) {
    cur_byte_ = uart_->read();
    if (state_ == 0) {
      if ((cur_byte_ == HEADER_) && ((prev_byte_ == FOOTER_) ||
         ((prev_byte_ & 0x0F) == FOOTER2_))) {
        buf_[state_++] = cur_byte_;
      } else {
        state_ = 0;
      }
    } else if (state_ < PAYLOAD_LEN_ + HEADER_LEN_) {
        buf_[state_++] = cur_byte_;
    } else if (state_ < PAYLOAD_LEN_ + HEADER_LEN_ + FOOTER_LEN_) {
      state_ = 0;
      prev_byte_ = cur_byte_;
      if ((cur_byte_ == FOOTER_) || ((cur_byte_ & 0x0F) == FOOTER2_)) {
        /* Grab the channel data */
        data_.ch[0]  = static_cast<int16_t>(buf_[1] | buf_[2] << 8 & 0x07FF);
        data_.ch[1]  = static_cast<int16_t>(buf_[2] >> 3 | buf_[3] << 5 &
                                            0x07FF);
        data_.ch[2]  = static_cast<int16_t>(buf_[3] >> 6 | buf_[4] << 2  |
                                            buf_[5] << 10 & 0x07FF);
        data_.ch[3]  = static_cast<int16_t>(buf_[5] >> 1 | buf_[6] << 7 &
                                            0x07FF);
        data_.ch[4]  = static_cast<int16_t>(buf_[6] >> 4 | buf_[7] << 4 &
                                            0x07FF);
        data_.ch[5]  = static_cast<int16_t>(buf_[7] >> 7 | buf_[8] << 1 |
                                            buf_[9] << 9 & 0x07FF);
        data_.ch[6]  = static_cast<int16_t>(buf_[9] >> 2 | buf_[10] << 6 &
                                            0x07FF);
        data_.ch[7]  = static_cast<int16_t>(buf_[10] >> 5 | buf_[11] << 3 &
                                            0x07FF);
        data_.ch[8]  = static_cast<int16_t>(buf_[12] | buf_[13] << 8 & 0x07FF);
        data_.ch[9]  = static_cast<int16_t>(buf_[13] >> 3 | buf_[14] << 5 &
                                            0x07FF);
        data_.ch[10] = static_cast<int16_t>(buf_[14] >> 6 | buf_[15] << 2 |
                                            buf_[16] << 10 & 0x07FF);
        data_.ch[11] = static_cast<int16_t>(buf_[16] >> 1 | buf_[17] << 7 &
                                            0x07FF);
        data_.ch[12] = static_cast<int16_t>(buf_[17] >> 4 | buf_[18] << 4 &
                                            0x07FF);
        data_.ch[13] = static_cast<int16_t>(buf_[18] >> 7 | buf_[19] << 1  |
                                            buf_[20] << 9 & 0x07FF);
        data_.ch[14] = static_cast<int16_t>(buf_[20] >> 2 | buf_[21] << 6 &
                                            0x07FF);
        data_.ch[15] = static_cast<int16_t>(buf_[21] >> 5 | buf_[22] << 3 &
                                            0x07FF);
        /* CH 17 */
        data_.ch17 = buf_[23] & CH17_MASK_;
        /* CH 18 */
        data_.ch18 = buf_[23] & CH18_MASK_;
        /* Grab the lost frame */
        data_.lost_frame = buf_[23] & LOST_FRAME_MASK_;
        /* Grab the failsafe */
        data_.failsafe = buf_[23] & FAILSAFE_MASK_;
        return true;
      } else {
        return false;
      }
    } else {
      state_ = 0;
    }
    prev_byte_ = cur_byte_;
  }
  return false;
}

/* Needed for emulating two stop bytes on Teensy 3.0 and 3.1/3.2 */
#if defined(__MK20DX128__) || defined(__MK20DX256__)
namespace {
  IntervalTimer serial_timer;
  HardwareSerial *sbus_bus;
  uint8_t sbus_packet[25];
  volatile int send_index = 0;
  void SendByte() {
    if (send_index < 25) {
      sbus_bus->write(sbus_packet[send_index]);
      send_index++;
    } else {
      serial_timer.end();
      send_index = 0;
    }
  }
}  // namespace
#endif

void SbusTx::Begin() {
  if (fast_) {
    baud_ = 200000;
  } else {
    baud_ = 100000;
  }
  /* Start the bus */
  /* Teensy 3.0 || Teensy 3.1/3.2 */
  #if defined(__MK20DX128__) || defined(__MK20DX256__)
  sbus_bus = uart_;
  if (inv_) {
    uart_->begin(baud_, SERIAL_8E1_RXINV_TXINV);
  } else {
    uart_->begin(baud_, SERIAL_8E1);
  }
  /*
  * Teensy 3.5 || Teensy 3.6 ||
  * Teensy LC  || Teensy 4.0/4.1 ||
  * Teensy 4.0 Beta
  */
  #elif defined(__MK64FX512__) || defined(__MK66FX1M0__) || \
        defined(__MKL26Z64__)  || defined(__IMXRT1062__) || \
        defined(__IMXRT1052__)
  if (inv_){
    uart_->begin(baud_, SERIAL_8E2_RXINV_TXINV);
  } else {
    uart_->begin(baud_, SERIAL_8E2);
  }
  /* STM32L4 */
  #elif defined(STM32L496xx) || defined(STM32L476xx) || \
        defined(STM32L433xx) || defined(STM32L432xx)
  if (inv_) {
    uart_->begin(baud_, SERIAL_8E2 | 0xC000ul);
  } else {
    uart_->begin(baud_, SERIAL_8E2);
  }
  /* ESP32 */
  #elif defined(ESP32)
  uart_->begin(baud_, SERIAL_8E2, rxpin_, txpin_, inv_);
  /* Everything else, with a hardware inverter */
  #else
  uart_->begin(baud_, SERIAL_8E2);
  #endif
}

void SbusTx::Write() {
  /* Assemble packet */
  buf_[0] = HEADER_;
  buf_[1] = static_cast<uint8_t>((data_.ch[0] & 0x07FF));
  buf_[2] = static_cast<uint8_t>((data_.ch[0] & 0x07FF) >> 8 |
            (data_.ch[1] & 0x07FF) << 3);
  buf_[3] = static_cast<uint8_t>((data_.ch[1] & 0x07FF) >> 5 |
            (data_.ch[2] & 0x07FF) << 6);
  buf_[4] = static_cast<uint8_t>((data_.ch[2] & 0x07FF) >> 2);
  buf_[5] = static_cast<uint8_t>((data_.ch[2] & 0x07FF) >> 10 |
            (data_.ch[3] & 0x07FF) << 1);
  buf_[6] = static_cast<uint8_t>((data_.ch[3] & 0x07FF) >> 7 |
            (data_.ch[4] & 0x07FF) << 4);
  buf_[7] = static_cast<uint8_t>((data_.ch[4] & 0x07FF) >> 4 |
            (data_.ch[5] & 0x07FF) << 7);
  buf_[8] = static_cast<uint8_t>((data_.ch[5] & 0x07FF) >> 1);
  buf_[9] = static_cast<uint8_t>((data_.ch[5] & 0x07FF) >> 9  |
            (data_.ch[6] & 0x07FF) << 2);
  buf_[10] = static_cast<uint8_t>((data_.ch[6] & 0x07FF) >> 6  |
             (data_.ch[7] & 0x07FF) << 5);
  buf_[11] = static_cast<uint8_t>((data_.ch[7] & 0x07FF) >> 3);
  buf_[12] = static_cast<uint8_t>((data_.ch[8] & 0x07FF));
  buf_[13] = static_cast<uint8_t>((data_.ch[8] & 0x07FF) >> 8 |
             (data_.ch[9]  & 0x07FF) << 3);
  buf_[14] = static_cast<uint8_t>((data_.ch[9] & 0x07FF) >> 5 |
             (data_.ch[10] & 0x07FF) << 6);
  buf_[15] = static_cast<uint8_t>((data_.ch[10] & 0x07FF) >> 2);
  buf_[16] = static_cast<uint8_t>((data_.ch[10] & 0x07FF) >> 10 |
             (data_.ch[11] & 0x07FF) << 1);
  buf_[17] = static_cast<uint8_t>((data_.ch[11] & 0x07FF) >> 7  |
             (data_.ch[12] & 0x07FF) << 4);
  buf_[18] = static_cast<uint8_t>((data_.ch[12] & 0x07FF) >> 4  |
             (data_.ch[13] & 0x07FF) << 7);
  buf_[19] = static_cast<uint8_t>((data_.ch[13] & 0x07FF) >> 1);
  buf_[20] = static_cast<uint8_t>((data_.ch[13] & 0x07FF) >> 9  |
             (data_.ch[14] & 0x07FF) << 2);
  buf_[21] = static_cast<uint8_t>((data_.ch[14] & 0x07FF) >> 6  |
             (data_.ch[15] & 0x07FF) << 5);
  buf_[22] = static_cast<uint8_t>((data_.ch[15] & 0x07FF) >> 3);
  buf_[23] = 0x00 | (data_.ch17 * CH17_MASK_) | (data_.ch18 * CH18_MASK_) |
             (data_.failsafe * FAILSAFE_MASK_) |
             (data_.lost_frame * LOST_FRAME_MASK_);
  buf_[24] = FOOTER_;
  /* Send packet to servos */
  #if defined(__MK20DX128__) || defined(__MK20DX256__)
  /* 
  * Use ISR to send byte at a time,
  * 130 us between bytes to emulate 2 stop bits
  */
  __disable_irq();
  memcpy(sbus_packet, buf_, sizeof(buf_));
  __enable_irq();
  serial_timer.priority(255);
  serial_timer.begin(SendByte, 130);
  #else
  uart_->write(buf_, sizeof(buf_));
  #endif
}

}  // namespace bfs
